### PR TITLE
Client-gen output

### PR DESCRIPTION
**Pull request by caesarxuchao**
This PR 1. changes the output directory of the client-gen according to the client package [proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/client-package-structure.md#package-structure); 2. generates the output.

I plan to make a follow-up PR that adds the uncommon methods of client (e.g. `func (c *pods) GetLogs(...)`, another PR to add the [clientset](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/client-package-structure.md#high-level-client-sets), and a final PR to replace `pkg/client/unversioned`.

First commit is #18231.
Commit "add OrderTypes() to Orderer" is #18687.
Commits "change output directory", "make the types ordered", and "output" are the important ones.

cc @krousey @wojtek-t

### CODE REVIEW

**Comment by wojtek-t** [comment]
@caesarxuchao - can you please add your script to:
this script:
hack/after-build/run-codegen.sh 
so that we will be running it on our regular tests?

**Comment by krousey** [comment]
So in general the output looks good. I'm still not a fan of the client interfaces, but that's not for this PR.

So this depends on the current unversioned client, and for good reason. RESTClient is still there. But also this uses the current client config, which has too many fields for our generated clients. You unconditionally set the prefix, the version, and the codec. Perhaps we should adopt a config where these are not options. Ideally reusing transport.Config, the only other fields it would need would be host, qps, and burst.

Now there's still some work to be done for that to get there, so I'm questioning if we want to check this in before that happens. I'm actually working on pulling the last bit of the version specific logic out of restclient today. I'll send you a PR when it's ready.

Also, the unversioned client is a bit special in that it does conversion, where the versioned typed clients won't do that (by default at least). Are we sure we want to replace it with auto-genned code?

**Comment by caesarxuchao** [comment]
> You unconditionally set the prefix, the version, and the codec. Perhaps we should adopt a config where these are not options. Ideally reusing transport.Config, the only other fields it would need would be host, qps, and burst.

In the client-package-structure proposal: "If you want to use a typed client against a server's API endpoint and the server's API version doesn't match the client's API version, you can construct the client with a RESTClient using a Codec that does the conversion (this is basically what our client does now)". So probably we should keep the fields like `APIVersion` and `Codec` here. WDYT?

**Comment by caesarxuchao** [comment]
> I'm actually working on pulling the last bit of the version specific logic out of restclient today. I'll send you a PR when it's ready.

Sure we can wait. I'm not clear how that will affect the typed client though. Seems this PR does not rely on any version specific behavior of RESTClient.

**Comment by caesarxuchao** [comment]
> Also, the unversioned client is a bit special in that it does conversion, where the versioned typed clients won't do that (by default at least).

I think the generated code should be the same for unversioned client and versioned client. The conversions happens at this line (and similar lines): https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/pods.go#L93
The `Body` and `Into` uses the codec to do the encoding/decoding. Whether they do any conversion depends on the codec, which is hidden from the generated client code.

**Comment by krousey** [comment]
>In the client-package-structure proposal: "If you want to use a typed client against a server's API endpoint and the server's API version doesn't match the client's API version, you can construct the client with a RESTClient using a Codec that does the conversion (this is basically what our client does now)". So probably we should keep the fields like APIVersion and Codec here. WDYT?

I definitely think these should be parameters to construct the rest client, but not the typed client. For the case you mention, I expect them to construct the rest client, and then just pass that in to a constructor. I imagine 2 constructors, one that just takes a restclient, and another that takes the stripped down config that I talked about.

> Sure we can wait. I'm not clear how that will affect the typed client though. Seems this PR does not rely on any version specific behavior of RESTClient.

This PR doesn't The version specific stuff is just gating me moving the restclient to its own package, which in turn will affect this PR.


> I think the generated code should be the same for unversioned client and versioned client. The conversions happens at this line (and similar lines): https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/pods.go#L93
The Body and Into uses the codec to do the encoding/decoding. Whether they do any conversion depends on the codec, which is hidden from the generated client code.

I do not. The codec that the unversioned client will pass into the rest client will be different, in that it will do conversions. With the typed clients, there should be no conversion in the generated code. This is something you would have to build a custom codec for and pass in the restclient. This will also be a slightly different interface (constructors, configs, ...) than what the current unversioned client provides, so there might be a migration cost. Also, I think we generally want to discourage and even eventually eliminate the unversioned client.

I'm probably wrong on all this, or at least they'll be some resistance. But I think it warrants a little bit of dicussion.

**Comment by caesarxuchao** [comment]
Link to the discussion on `// +genclient=true`:
https://github.com/caesarxuchao/kubernetes/commit/c0e76818691896a9e15cef863cdffe33df15604b#commitcomment-14983270

**Comment by caesarxuchao** [comment]
> I definitely think these should be parameters to construct the rest client, but not the typed client. For the case you mention, I expect them to construct the rest client, and then just pass that in to a constructor. I imagine 2 constructors, one that just takes a restclient, and another that takes the stripped down config that I talked about.

Ok. I agree with this part. Just a little concern that making the stripped down config another struct might be over complicated. I'll make the changes in this PR.

**Comment by caesarxuchao** [comment]
> I do not. The codec that the unversioned client will pass into the rest client will be different, in that it will do conversions. 

I agree these code will be different for unversioned client and versioned client: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/extensions.go#L152-L157

> With the typed clients, there should be no conversion in the generated code. This is something you would have to build a custom codec for and pass in the restclient.

smarterclayton is creating serializer in #17922. We probably will be able to get a serializer from throught `latest` without the need of "build a custom codec".

> This will also be a slightly different interface (constructors, configs, ...) than what the current unversioned client provides, so there might be a migration cost.

I disagree. If we implement the two constructors as you suggested, one that takes RESTClient and another one that take the stripped down config, the constructors should be the same for unversioned and versioned clients.

> Also, I think we generally want to discourage and even eventually eliminate the unversioned client.

Yes.

I agree I might need to make some changes to the client-gen when we generate versioned client. I still want to check in the code for the unversioned client, because I want to test the generated client in production and I don't see we can migrate our code base to use the versioned client soon enough :)

File: pkg/api/types.go [hunk 1]
```diff
@@ -310,6 +312,8 @@ type PersistentVolumeList struct {
 	Items                []PersistentVolume `json:"items"`
 }
 
+// +genclient=true
+
 // PersistentVolumeClaim is a user's request for and claim to a persistent volume
 type PersistentVolumeClaim struct {
```
Comment (caesarxuchao) [inline pkg/api/types.go]: For client-gen,  the tag must be in the comment before the type's doc. This prevents the tag from showing up in godoc or our API docs (definitions.html).

Users of go2idl can choose whether the tag should exist in the immediate comment block or the comment block before it.

File: pkg/client/typed/generated/legacy/unversioned/legacy_client.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+func NewLegacyForConfigOrDie(c *unversioned.Config) *LegacyClient {
+	client, err := NewLegacyForConfig(c)
+	if err != nil {
+		panic(err)
+	}
+	return client
+}
+
+// NewLegacy creates a new LegacyClient for the given RESTClient.
+func NewLegacyForRESTClient(c *unversioned.RESTClient) *LegacyClient {
```
Comment (caesarxuchao) [inline pkg/client/typed/generated/legacy/unversioned/legacy_client.go]: @krousey this is the second constructor we talked about.

File: pkg/client/typed/generated/legacy/unversioned/legacy_client.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+func (c *LegacyClient) Services(namespace string) ServiceInterface {
+	return newServices(c, namespace)
+}
+
+func (c *LegacyClient) ServiceAccounts(namespace string) ServiceAccountInterface {
+	return newServiceAccounts(c, namespace)
+}
+
+// NewLegacy creates a new LegacyClient for the given config.
+func NewLegacyForConfig(c *unversioned.Config) (*LegacyClient, error) {
```
Comment (caesarxuchao) [inline pkg/client/typed/generated/legacy/unversioned/legacy_client.go]: @krousey, can we make the strip down config in another PR? This one is quite large already.

**Comment by caesarxuchao** [comment]
@krousey PTAL. Let me know if you want to make more changes in this PR. Thanks!

File: cmd/libs/go2idl/client-gen/generators/generator-for-group.go [hunk 4]
```diff
@@ -109,9 +117,9 @@ func (c *$.Group$Client) $.type|publicPlural$(namespace string) $.type.Name.Name
 }
 `
 
-var newClientTemplate = `
+var newClientForConfigTemplate = `
 // New$.Group$ creates a new $.Group$Client for the given config.
-func New$.Group$(c *$.Config|raw$) (*$.Group$Client, error) {
+func New$.Group$ForConfig(c *$.Config|raw$) (*$.Group$Client, error) {
```
Comment (krousey) [inline cmd/libs/go2idl/client-gen/generators/generator-for-group.go]: Doc string no longer matches

File: cmd/libs/go2idl/client-gen/generators/generator-for-group.go [hunk 5]
```diff
@@ -124,11 +132,11 @@ func New$.Group$(c *$.Config|raw$) (*$.Group$Client, error) {
 }
 `
 
-var newClientOrDieTemplate = `
+var newClientForConfigOrDieTemplate = `
 // New$.Group$OrDie creates a new $.Group$Client for the given config and
 // panics if there is an error in the config.
-func New$.Group$OrDie(c *$.Config|raw$) *$.Group$Client {
-	client, err := New$.Group$(c)
+func New$.Group$ForConfigOrDie(c *$.Config|raw$) *$.Group$Client {
```
Comment (krousey) [inline cmd/libs/go2idl/client-gen/generators/generator-for-group.go]: doc string

File: cmd/libs/go2idl/client-gen/generators/generator-for-group.go [hunk 6]
```diff
@@ -136,10 +144,16 @@ func New$.Group$OrDie(c *$.Config|raw$) *$.Group$Client {
 }
 `
 
+var newClientForRESTClientTemplate = `
+// New$.Group$ creates a new $.Group$Client for the given RESTClient.
+func New$.Group$ForRESTClient(c *$.RESTClient|raw$) *$.Group$Client {
```
Comment (krousey) [inline cmd/libs/go2idl/client-gen/generators/generator-for-group.go]: doc string. And perhaps this should just be called New?

**Comment by krousey** [comment]
Couple of really minor comments. Also, why put the group name in the constructors? And is legacy a group? Is that going to change?

File: cmd/libs/go2idl/client-gen/generators/generator-for-group.go (hunk 6 above)
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/generator-for-group.go]: Right, the group name will be in the imported package name.

**Comment by caesarxuchao** [comment]
> And is legacy a group?

No, it's not. We usually call it "". Calling it "" is kind of annoying, that will break the directory structure, and make names confusing. In lavalamp's proposal, it's called the "API" group (https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/client-package-structure.md#high-level-client-sets), but I think that name is also confusing. So far I think "legacy" is our best choice. What do you think?

cc @nikhiljindal

**Comment by krousey** [comment]
Would "core" be better?

**Comment by caesarxuchao** [comment]
I'd like to reserve the word "core" because we might actually have a separate "core" group in the future, it will include APIs like Pod.

File: cmd/libs/go2idl/client-gen/generators/generator-for-group.go (hunk 4 above)
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/generator-for-group.go]: Done

File: cmd/libs/go2idl/client-gen/generators/generator-for-group.go (hunk 5 above)
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/generator-for-group.go]: Done. And the function name changed to NewForConfigOrDie.

File: cmd/libs/go2idl/client-gen/generators/generator-for-group.go (hunk 6 above)
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/generator-for-group.go]: Done. The function name is changed to New.

**Comment by caesarxuchao** [comment]
@krousey, so far the only unresolved issue is the group name "legacy", I prefer it over "core", but I'm open to other options :)

**Comment by krousey** [comment]
Legacy is fine for now. If you fixed the names, feel free to apply the lgtm

**Comment by caesarxuchao** [comment]
Yeah, the names are fixed. I'm applying the label. Yay!

Thanks.

**Comment by caesarxuchao** [comment]
squashed and added the lgtm label.

**Comment by caesarxuchao** [comment]
This doesn't affect our production code, so added the e2e-not-required label.
